Add CRD versioning developer doc (#2404)
nginxinc/nginx-gateway-fabric: docs/developer/crd-versioning.md

# CRD Versioning

This document discusses how we version our CustomResourceDefinitions (CRDs) that define Policies, Filters, or other configuration objects. NGINX Gateway Fabric distributes a few CRDs, such as NginxGateway, NginxProxy, ClientSettingsPolicy, and ObservabilityPolicy. More will inevitably be added over time.

## Initial Version

When creating a new CRD, the initial version should be `v1alpha1`. This version indicates a new API that needs time for use and feedback. By starting at the alpha version, it also allows us a little more flexibility in changing the API.

## Changing an API

When changing an API in a CRD, it's important to understand the impact of those changes. This will determine if and how we need to update the API version.

**No change in version required**

> - Adding new optional fields.

Reviewer: Let's add documentation updates to this list

> **Compatible changes, requires a version change**

Reviewer: Can we add a link to the k8s doc on compatible changes somewhere in this section: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility?

We'll allow a bit more flexibility for this case when dealing with alpha APIs.

- Adding new required fields.
- Removing fields. If possible, we should try to deprecate fields for 3 releases to slowly phase them out before removing.
- Changing validation and allowed values of a field.

For more in depth information on how to handle API changes, see the Kubernetes [API changes doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md).

NGF version should not need to change when CRD versions change. CRDs are not embedded into NGF; they have their own versions. NGF releases are directly tied to the CRD version they support and users are required to upgrade CRD versions when upgrading NGF. This ensures that the CRD version is supported by NGF. However, the user may need to fix their CRDs if they aren't compatible.

> ## Graduating to v1 (stable)

Reviewer: Why are we jumping right from alpha to v1?

Author: This was following the Gateway API pattern. We certainly don't have to, but are our APIs going to be worth going through 3 stages of API versioning?

Reviewer: Probably not. Can we add a note in this doc explaining why we are skipping beta?

# CRD Versioning

> This document discusses how we version our CustomResourceDefinitions (CRDs) that define Policies, Filters, or other configuration objects. NGINX Gateway Fabric distributes a few CRDs, such as NginxGateway, NginxProxy, ClientSettingsPolicy, and ObservabilityPolicy. More will inevitably be added over time.

Reviewer 2: ```suggestion
This document discusses how we version our CustomResourceDefinitions (CRDs) that define Policies, Filters, or other configuration objects. 

NGINX Gateway Fabric currently distributes the following CRDs: 

- NginxGateway
- NginxProxy,
- ClientSettingsPolicy
- ObservabilityPolicy.
```
If the list is going to get bigger in the future, swapping it to be bullets will help with readability.

Best to stick to statements about the current version of the project without making promises about the future.

Author: My hope was that we don't want to have to update this doc every time we add a new CRD, and just supply a sample list of some that we distribute.

For more in depth information on compatible changes, see the Kubernetes [API changes doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility).

**Breaking changes, requires a version change**

The following API changes are incompatible with previous API versions, and therefore not only require a version bump, but cannot use a conversion webhook. These types of changes should be avoided if at all possible due to the disruption for users. A user will need to update their configurations when upgrading NGF. These types of changes need clear messaging in release notes and docs.

We'll allow a bit more flexibility for this case when dealing with alpha APIs.

- Adding new required fields.
- Removing fields. If possible, we should try to deprecate fields for 3 releases to slowly phase them out before removing.
- Changing validation and allowed values of a field.

For more in depth information on how to handle API changes, see the Kubernetes [API changes doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas).

> 
> NGF version should not need to change when CRD versions change. CRDs are not embedded into NGF; they have their own versions. NGF releases are directly tied to the CRD version they support and users are required to upgrade CRD versions when upgrading NGF. This ensures that the CRD version is supported by NGF. However, the user may need to fix their CRDs if they aren't compatible.

Reviewer 2: ```suggestion
The NGINX Gateway Fabric version should not need to change when CRD versions change. CRDs are not embedded into NGF; they have their own versions. 

NGINX Gateway Fabric releases are directly tied to the CRD version they support and users are required to upgrade CRD versions when upgrading NGF. This ensures that the CRDs remain compatible with NGINX Gateway Fabric.

However, the user may need to fix any incompatible CRDs.
```

Avoiding three letter acronyms, spacing, phrasing.

## Graduating to v1 (stable)

> Having an alpha API allows us more flexibility in making breaking API changes as we further design and craft the API to fit the necessary use cases. However, at some point these APIs need to become stable and no longer have breaking changes, so that a user can rely on them without the worry of things not working from release to release.

Reviewer 2: ```suggestion
An alpha API allows us more flexibility in making breaking API changes as we further design and craft the API to fit the necessary use cases. However, at some point these APIs need to become stable and no longer have breaking changes, so that a user can rely on them without the worry of things not working from release to release.
```
